lumicks/pylake/detail/tests/test_value.py

```diff
@@ -0,0 +1,183 @@
+import math
+import operator
+
+import numpy as np
+import pytest
+
+from lumicks.pylake.detail.value import ValueMixin
+
+
+class Parameter(ValueMixin):
+    def __init__(self, value, description):
+        self.value = value
+        self.description = description
+
+
+class BadParameter(ValueMixin):
+    def __init__(self, description):
+        self.description = description
+
+
+operators = {
+    operator.add,
+    operator.sub,
+    operator.mul,
+    operator.truediv,
+    operator.floordiv,
+    operator.mod,
+    divmod,
+    operator.pow,
+    operator.and_,
+    operator.or_,
+    operator.xor,
+}
+
+
+def test_mixin():
+    a = Parameter(2, "two")
+    a + 2
+
+    b = BadParameter("two")
+    with pytest.raises(AttributeError, match="'BadParameter' object has no attribute 'value'"):
+        b + 2
+
+
+def test_operators():
+    a = Parameter(a0 := 5, "five")
+    b = Parameter(b0 := 8, "eight")
+
+    for op in operators:
+        assert op(a0, 2) == op(a, 2)
+        assert op(2, a0) == op(2, a)
+        assert op(a0, b0) == op(a, b)
+        assert op(b0, a0) == op(b, a)
+
+
+def test_comparisons():
+    a = Parameter(5, "five")
+
+    assert a == 5
+    assert 5 == a
+
+    assert a != 4
+    assert 4 != a
+
+    assert a < 10
+    assert 10 > a
+
+    assert a > 2
+    assert 2 < a
+
+    assert a >= 5
+    assert 5 >= a
+    assert a <= 5
+    assert 5 <= a
+
+
+def test_operators_str():
+    a = Parameter(a0 := "a", "alpha")
+    b = Parameter(b0 := "b", "beta")
+
+    assert a0 + "c" == a + "c"
+    assert "c" + a0 == "c" + a
+    assert a0 + b0 == a + b
+    assert a0 * 5 == a * 5
+    assert a0 * 5 == a * Parameter(5, "five")
+    assert a0 * Parameter(5, "five") == a * 5
+
+    assert a == "a"
+
+    with pytest.raises(TypeError, match="not all arguments converted during string formatting"):
+        a % 1
+
+    assert "Hello, venus" == Parameter("Hello, %s", "fmt") % "venus"
+
+    for c in (1, "c", True):
+        for op in operators - {operator.add, operator.mul, operator.mod}:
+            with pytest.raises(TypeError):
+                op(a, c)
+
+
+def test_operators_bool():
```

Comment:
I think you're missing the cast to bool, which is pretty important I think as it is required for a lot of control flow.

The following example fails right now for instance:
```python
    a = Parameter(None, "five")
    assert not a
```
And that seems a particularly risky one, since it might be checked to see whether something is defined or not.

There's a second gotcha, where I don't think you can get around as easily which is the check:

```python
assert a is None
```
I know that one is used frequently to see if something is set at all, and that may be one where the abstraction may be dicey since it can never evaluate to true, since it isn't the same entity.

I have to think about this one.

Reply:
yeah this is kinda tricky, but on the otherhand the intent is to be able to use a class with a numeric `value` as if it were a scalar. So I'm not too worried about emulating all possible functionality for any possible type in `value`.

That being said, I've added explicit tests
```
Parameter(None) == None  # passes
Parameter(None) is None  # fails

Reply:
This particular one concerns me because the "false positive" (`None` claiming it is not `None`) could lead to subtle hard to spot bugs.

We use `None` as a signaling value for unspecified values for optionals and many other packages do too. Because just doing `if (var)` would also be `False` for zero, which is often not what you intend.

We could enforce to be explicit on our end (using the equality operator), but other packages are not under our control and as far as I know `a is None` is considered idiomatic and less ambiguous, as it doesn't invoke the equality operator which could be overloaded.